Visual enhancements to TestRunner

mrrobinhood5 opened the pull request:
- Option to display description and examples
- Used tables to display result
- added some colors

mrrobinhood5 committed:
- Option to display description and examples
- Used tables to display result
- added some colors

ZechCodes commented on beginnercodes/challenges/test_runner.py:
```diff
@@ -23,17 +29,41 @@ class Result:
+    return _t
+
+
+def test(challenge: int, solution_func, description: bool = False, examples: bool = False):
```
Make `solution_func` optional, no need to pass in a function if you just wanna get the description or examples

ZechCodes commented on beginnercodes/challenges/test_runner.py:
```diff
@@ -23,17 +29,41 @@ class Result:
-        if response.status_code == 404:
-            raise ChallengeNotFound(f"Challenge {challenge} was not found.")
+    def _fetch_data(url:str) -> Union[Response, str]:
+        _r = requests.get(url.replace('XXXX', str(challenge)))
```
Don't use `replace`, use an f-string (the way it was before)

ZechCodes commented on pyproject.toml:
The package version at the top of the file needs to be bumped to 0.1.2

ZechCodes commented on beginnercodes/challenges/test_runner.py:
```diff
@@ -23,17 +29,41 @@ class Result:
-        )
-        if response.status_code == 404:
-            raise ChallengeNotFound(f"Challenge {challenge} was not found.")
+    def _fetch_data(url:str) -> Union[Response, str]:
```
`fetch_data` never returns a string only a `Response`, so the union is unnecessary. This would remove the need for the `Union` import from `typing`.

ZechCodes commented on beginnercodes/challenges/test_runner.py:
```diff
@@ -23,17 +29,41 @@ class Result:
+        return _r
 
+    def _get_tests(challenge: int) -> list[dict]:
+        url = "https://raw.githubusercontent.com/beginner-codes/challenges/main/weekday/test_cases_XXXX.json"
```
This should be an f-string instead of using `XXXX`

ZechCodes commented on beginnercodes/challenges/test_runner.py:
```diff
@@ -23,17 +29,41 @@ class Result:
         return response.json()
 
+    def _get_info(challenge: int, description: bool, examples: bool) -> str:
+        url = "https://raw.githubusercontent.com/beginner-codes/challenges/main/weekday/challenge_XXXX.md"
```
This should be an f-string instead of using `XXXX`

ZechCodes commented on beginnercodes/challenges/test_runner.py:
```diff
@@ -23,17 +29,41 @@ class Result:
+            info = _fetch_data(url)
+            try:
+                result += info.text.split('##')[0] if description else ''
+                result += '# ' + info.text.split('# ')[2].replace('#','') if examples else ''
```
Don't use a ternary if's here. Use `if description` and `elif examples`. Simpler, easier to understand, and results in less op codes to be run.

mrrobinhood5 committed:
- bumped version in toml file
- passing a function is now optional
- passing incorrect arguments prints USAGE
- _fetch_data only returns Response (No Union)
- fixed usages of .replace
- cleared up _get_info instead of terniary ifs

mrrobinhood5 commented:
Added a commit fixing all them things and added 1 more thing in case no function/examples/description args are passed.

mrrobinhood5 committed:
- fixed a minor bug with USAge

ZechCodes commented:
Apologies for the extreme delay, all looks good, I’ll merge the changes. Thanks!
